src/backend/optimizer/plan/aqumv.c

```diff
@@ -0,0 +1,796 @@
+		aqumv_adjust_varno(mvQuery, 1);
+
+		/*
+		 * AQUMV_FIXME_MVP
+		 * Are stable functions OK?
+		 * A STABLE function cannot modify the database and is guaranteed to
+		 * return the same results given the same arguments for all rows
+		 * within a single statement.
+		 * But AQUMV rewrites the query to a new SQL actually, though the same
+		 * results is guaranteed.
+		 * Its's unclear whether STABLE is OK, let's be conservative for now.
+		 */
+		if(contain_mutable_functions((Node *)mvQuery))
+			continue;
+
+		context = aqumv_init_context(mvQuery->targetList, mvRelQueryTree->targetList);
+
+		/* Sort nonpure vars expression, prepare for Greedy Algorithm. */
+		aqumv_process_nonpure_vars_expr(context);
+
+		/*
+		 * Process and rewrite target list, return false if failed.
+		 */
+		if(!aqumv_process_targetlist(context, parse->targetList, &mv_final_tlist))
+			continue;
+
+		/*
+		 * AQUMV
+		 * Process all quals to conjunctive normal form.
+		 * 
+		 * We assume that the selection predicates of view and query expressions
+		 * have been converted into conjunctive normal form(CNF) before we process
+		 * them.
+		 * AQUMV_MVP: no having quals now.
+		 */
+		preprocess_qual_conditions(subroot, (Node *) mvQuery->jointree);
+
+		/*
+		 * Process quals, return false if failed. 
+		 * Else, post_quals are filled if there were. 
+		 * Like process target list, post_quals is used later to see if we could
+		 * rewrite and apply it to mv relation.
+		 */
+		if(!aqumv_process_from_quals(parse->jointree->quals, mvQuery->jointree->quals, &post_quals))
+			continue;
+
+		/* Rewrite post_quals, return false if failed. */
+		post_quals = (List *)aqumv_adjust_sub_matched_expr_mutator((Node *)post_quals, context);
+		if (context->has_unmatched)
+			continue;
+
+		/*
+		 * Here! We succeed to rewrite a new SQL.
+		 * Begin to replace all guts.
+		 */
+		mvQuery->targetList = mv_final_tlist;
+
+		/*
+		 * AQUMV
+		 * NB: Update processed_tlist again in case that tlist has been changed. 
+		 */
+		preprocess_targetlist(subroot);
+
+		/*
+		 * AQUMV
+		 * NB: Correct the final_locus as we select from another realtion now.
+		 */
+		PathTarget *newtarget = make_pathtarget_from_tlist(subroot->processed_tlist);
+		subroot->final_locus = cdbllize_get_final_locus(subroot, newtarget);
+
+		/* Rewrite with mv's query tree*/
+		mvrte->relkind = RELKIND_MATVIEW;
+		mvrte->relid = matviewRel->rd_rel->oid;
+		/*
+		 * AQUMV_FIXME_MVP
+		 * Not sure where it's true from actions even it's not inherit tables.
+		 */
+		mvrte->inh = false;
+		mvQuery->rtable = list_make1(mvrte); /* rewrite to SELECT FROM mv itself. */
+		mvQuery->jointree->quals = (Node *)post_quals; /* Could be NULL, but doesn'y matter for now. */
+
+		/*
+		 * AQUMV
+		 * Build a plan of new SQL.
+		 * AQUMV is cost-based, let planner decide which is better.
+		 * AQUMV_FIXME_MVP:
+		 * 	no qp_callback function now.
+		 * 	replcace one-by-one?
+		 */
+		mv_final_rel = query_planner(subroot, NULL, NULL);
```

Comment:
why no callback ?

Reply:
Hi, thanks for review.
AQUMV don't  support window/group/distinct clause now. We don't need a callback function to compute pathkeys for them.